### ISSUE TITLE
Unnecessary high memory usage prevents using larger batch sizes.

### ISSUE THREAD

**Issue by XapaJIaMnu** [marian-nmt/marian-train#86]
Hey,

I have been doing various experiments with distributed and single node training and I noticed the following:
Using the default settings from the _examples/training_ I am able to reach about 38000 WPS (with gradient dropping) during training on 8x Pascal Titan X. However I noticed that I was using just 4461 MB of the 12 GBs available on the GPU.

Now if I change a few things (-w 7000 --mini-batch 512 --maxi-batch 768, up from 300, 64 and 100 respectively) ) I am able to utilize a lot more memory (11137 MB) during training and reach **45500**  WPS (averaged over a whole epoch).

However there is a crash during evaluation:
```
[2017-07-10 07:34:37] Ep. 1 : Up. 9000 : Sen. 2289954 : Cost 65.95 : Time 125.54s : 50420.62 words/s                                                
[2017-07-10 07:35:29] Seen 2395954 samples                                
[2017-07-10 07:35:29] Starting epoch 2                                    
[2017-07-10 07:35:29] [data] Shuffling files                              
[2017-07-10 07:35:40] [data] Done                                         
[2017-07-10 07:37:10] Ep. 2 : Up. 10000 : Sen. 147799 : Cost 62.97 : Time 153.75s : 41551.93 words/s                                                
[2017-07-10 07:37:11] Saving model to model/model.iter10000.npz           
[2017-07-10 07:37:12] Saving model to model/model.npz                     
[2017-07-10 07:37:13] [memory] Extending reserved space to 7680 MB (device 3)                                                                       
[2017-07-10 07:37:17] [memory] Extending reserved space to 14848 MB (device 3)                                                                      
terminate called after throwing an instance of 'thrust::system::system_error'                                                                       
terminate called recursively                                              
terminate called recursively                                              
  what():  after cub_::DeviceRadixSort::SortKeys(0): an illegal memory access was encountered
```
@tomekd suggests it's related to the computational graph being loaded a second time to calculate cross entropy. It should really be possible to swap out the currently unused GPU memory to RAM and then return it so that we can fully utilize the available GPU memory. This is not only a problem of the batch size, this sets an artificial limit of the size of the models trained.

Cheers,

Nick

**Comment by emjotde** [comment]
Nope. This is not what is happening (my guess). What's more likely is that the validation set contains sentences that are longer than the training set sentences, so it needs more memory for the forward step. The validation step is not using any additional memory by itself (at least that's what I remember, it should be reusing all graph instances already).

However, this is what --dynamic-batching is for. Set the memory as high as possible and do not worry about the rest. Leave mini-batch at the default, because dynamic-batching is currently not being used during validation, so choose a smaller batch size to avoid the above issue.

**Comment by emjotde** [comment]
BTW. maxi-batch size is just a muliplier for how many batches should be preloaded and sorted together from the corpus, it does not affect batch-size, it may only how a batch is filled.

**Comment by emjotde** [comment]
And another comment, higher WPS is not necessarily better here. Take into account that larger batch size may very well mean slower convergence.

**Comment by XapaJIaMnu** [comment]
Yes, I am aware of the slower convergence issue. I'm trying to find a sweet spot. 

I found that I can get better performance without hurting convergence with --mini-batch 128. With --mini-batch 192 (both of them with dynamic batching), i get even better performance with convergence delayed by one epoch, resulting in the same training time.

Also I found that when --dynamic-batching is omitted, --mini-batch and --maxi-batch are ignored (or at the very least they revert to the default according to the LOG info)

**Comment by emjotde** [comment]
Yes, mini-batch is ignored with dynamic-batching as it should be, maxi-batch is not being ignored, or it should not. Can you verify this?

The sentence "With --mini-batch 192 (both of them with dynamic batching)" makes no sense? After all-mini-batch is ignored with dynamic batching (only used for validation)

**Comment by XapaJIaMnu** [comment]
I think we misunderstood each other: This is with dynamic batching and manually specifying minibatch. I don't think it's ignored:

```
# train model
if [ ! -e "model/model.npz" ]
then                         

    ../../../marian/build/marian \
        --model model/model.npz \ 
        --devices $GPUS --seed 0 \
        --train-sets data/corpus.bpe.ro data/corpus.bpe.en \
        --vocabs model/vocab.ro.yml model/vocab.en.yml \    
        --dim-vocabs 66000 50000 \                          
        --dynamic-batching -w 3000 \                        
        --mini-batch 512 --maxi-batch 1024 \                
        --early-stopping 5 \                                
        --valid-freq 10000 --save-freq 10000 --disp-freq 1000 \
        --valid-sets data/newsdev2016.bpe.ro data/newsdev2016.bpe.en \
        --valid-metrics cross-entropy valid-script \                  
        --valid-script-path ./scripts/validate.sh \                   
        --log model/train.log --valid-log model/valid.log             

fi
[2017-07-10 09:05:02] [config] after-batches: 0
[2017-07-10 09:05:02] [config] after-epochs: 0 
[2017-07-10 09:05:02] [config] cell-dec: gru   
[2017-07-10 09:05:02] [config] cell-enc: gru   
[2017-07-10 09:05:02] [config] clip-norm: 1    
[2017-07-10 09:05:02] [config] devices:        
[2017-07-10 09:05:02] [config]   - 0           
[2017-07-10 09:05:02] [config] dim-emb: 512    
[2017-07-10 09:05:02] [config] dim-pos: 0      
[2017-07-10 09:05:02] [config] dim-rnn: 1024   
[2017-07-10 09:05:02] [config] dim-vocabs:     
[2017-07-10 09:05:02] [config]   - 66000       
[2017-07-10 09:05:02] [config]   - 50000       
[2017-07-10 09:05:02] [config] disp-freq: 1000 
[2017-07-10 09:05:02] [config] drop-rate: 0    
[2017-07-10 09:05:02] [config] dropout-rnn: 0  
[2017-07-10 09:05:02] [config] dropout-src: 0  
[2017-07-10 09:05:02] [config] dropout-trg: 0  
[2017-07-10 09:05:02] [config] dynamic-batching: true
[2017-07-10 09:05:02] [config] early-stopping: 5     
[2017-07-10 09:05:02] [config] guided-alignment-cost: ce
[2017-07-10 09:05:02] [config] guided-alignment-weight: 1
[2017-07-10 09:05:02] [config] keep-best: false          
[2017-07-10 09:05:02] [config] layer-normalization: false
[2017-07-10 09:05:02] [config] layers-dec: 1             
[2017-07-10 09:05:02] [config] layers-enc: 1             
[2017-07-10 09:05:02] [config] learn-rate: 0.0001        
[2017-07-10 09:05:02] [config] log: model/train.log      
[2017-07-10 09:05:02] [config] lr-decay: 0.7             
[2017-07-10 09:05:02] [config] lr-decay-freq: 50000      
[2017-07-10 09:05:02] [config] lr-decay-start:
[2017-07-10 09:05:02] [config]   - 10
[2017-07-10 09:05:02] [config]   - 1
[2017-07-10 09:05:02] [config] lr-decay-strategy: epoch+stalled
[2017-07-10 09:05:02] [config] max-length: 50
[2017-07-10 09:05:02] [config] maxi-batch: 1024
[2017-07-10 09:05:02] [config] mini-batch: 512
[2017-07-10 09:05:02] [config] mini-batch-words: 0
[2017-07-10 09:05:02] [config] model: model/model.npz
[2017-07-10 09:05:02] [config] moving-average: false
[2017-07-10 09:05:02] [config] moving-decay: 0.999
```

This is **without** dynamic batching and manually specifying --mini-batch and it IS ignored:

```
# train model
if [ ! -e "model/model.npz" ]
then                         

    ../../../marian/build/marian \
        --model model/model.npz \ 
        --devices $GPUS --seed 0 \
        --train-sets data/corpus.bpe.ro data/corpus.bpe.en \
        --vocabs model/vocab.ro.yml model/vocab.en.yml \    
        --dim-vocabs 66000 50000 \                          
#        --dynamic-batching -w 3000 \                       
        --mini-batch 512 --maxi-batch 1024 \                
        --early-stopping 5 \                                
        --valid-freq 10000 --save-freq 10000 --disp-freq 1000 \
        --valid-sets data/newsdev2016.bpe.ro data/newsdev2016.bpe.en \
        --valid-metrics cross-entropy valid-script \                  
        --valid-script-path ./scripts/validate.sh \                   
        --log model/train.log --valid-log model/valid.log             

fi
[2017-07-10 09:06:34] [config] after-batches: 0
[2017-07-10 09:06:34] [config] after-epochs: 0 
[2017-07-10 09:06:34] [config] cell-dec: gru   
[2017-07-10 09:06:34] [config] cell-enc: gru   
[2017-07-10 09:06:34] [config] clip-norm: 1    
[2017-07-10 09:06:34] [config] devices:        
[2017-07-10 09:06:34] [config]   - 0           
[2017-07-10 09:06:34] [config] dim-emb: 512    
[2017-07-10 09:06:34] [config] dim-pos: 0      
[2017-07-10 09:06:34] [config] dim-rnn: 1024   
[2017-07-10 09:06:34] [config] dim-vocabs:     
[2017-07-10 09:06:34] [config]   - 66000       
[2017-07-10 09:06:34] [config]   - 50000       
[2017-07-10 09:06:34] [config] disp-freq: 1000 
[2017-07-10 09:06:34] [config] drop-rate: 0    
[2017-07-10 09:06:34] [config] dropout-rnn: 0  
[2017-07-10 09:06:34] [config] dropout-src: 0  
[2017-07-10 09:06:34] [config] dropout-trg: 0  
[2017-07-10 09:06:34] [config] dynamic-batching: false
[2017-07-10 09:06:34] [config] early-stopping: 10     
[2017-07-10 09:06:34] [config] guided-alignment-cost: ce
[2017-07-10 09:06:34] [config] guided-alignment-weight: 1
[2017-07-10 09:06:34] [config] keep-best: false          
[2017-07-10 09:06:34] [config] layer-normalization: false
[2017-07-10 09:06:34] [config] layers-dec: 1
[2017-07-10 09:06:34] [config] layers-enc: 1
[2017-07-10 09:06:34] [config] learn-rate: 0.0001
[2017-07-10 09:06:34] [config] lr-decay: 0.7
[2017-07-10 09:06:34] [config] lr-decay-freq: 50000
[2017-07-10 09:06:34] [config] lr-decay-start:
[2017-07-10 09:06:34] [config]   - 10
[2017-07-10 09:06:34] [config]   - 1
[2017-07-10 09:06:34] [config] lr-decay-strategy: epoch+stalled
[2017-07-10 09:06:34] [config] max-length: 50
[2017-07-10 09:06:34] [config] maxi-batch: 100
[2017-07-10 09:06:34] [config] mini-batch: 64
[2017-07-10 09:06:34] [config] mini-batch-words: 0
```

I'm not sure if this is just an artifact from the logging, I spent a bit of time tracing it but I didn't look further...

**Comment by emjotde** [comment]
Ok. So what we actually need here is a second option, something like `--mini-batch-valid 64` to set the validation batch size independently.

**Comment by XapaJIaMnu** [comment]
So is the argument being ignored the second case or...?

**Comment by emjotde** [comment]
--mini-batch will not affect actual mini-batch size if --dynamic-batching is used, it will however be used for validation and during calculating the amount of sentences to preload, as this is the product of mini-batch and maxi-batch. 

Since with larger mini-batch the number of preloaded sentences increases it can affect training speed, because more sentences means better sorting according to target length which can increase WPS. 

The issue reported here is caused by the fact that the large mini-batch meets long sentences that exceed the capped sentence length during training. There is no cap for validation as it should validate on the full validation set. Hence a second mini-batch option would solve that.

**Comment by XapaJIaMnu** [comment]
Ok, but why is --mini-batch not recognized when used by itself?

**Comment by emjotde** [comment]
Is there a `model.npz.yml` file in the folder?

**Comment by emjotde** [comment]
Wait... that's a bash error! Everything after the comment is ignored, isn't it?

**Comment by XapaJIaMnu** [comment]
Ugh, yeah, my bad... I thought bash didn't have multiline comments...

**Comment by emjotde** [comment]
Anyway, having `--mini-batch-valid 64` probably makes sense.

**Comment by XapaJIaMnu** [comment]
Definitely. It would certainly make things less confusing and it'd make it easy to do convergence experiments with batch size growth.

**Comment by snukky** [comment]
Option `--valid-mini-batch` is added with 713865c6ff36c20fafba7fc71f33dd4e373f1aad. It forces the mini-batch size, so that `--mini-batch-words` and `--dynamic-batching` are not used for validation any more. It's consistent with how translation and rescoring work.

**Comment by emjotde** [comment]
@XapaJIaMnu Can you test this?

**Comment by XapaJIaMnu** [comment]
Yes, sorry for the delay. It seems to be working now.

Cheers,

Nick

**Comment by emjotde** [comment]
Great, actually just had to use it myself. Thanks, @snukky
